### ISSUE TITLE
job-manager: synchronize priority updates with job-info, etc

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#3342]
Peeling a discussion off of #3311 (priority plugin).

The PRIORITY state PR just merged to RFC 21 has a couple problems surrounding instance restart:

1) The `priority` event, which transitions PRIORITY to SCHED is a JOURNAL_ONLY event, so that when the instance restarts, the state pops back to PRIORITY.  That cycle is desirable but the fact that no event triggers the retrograde transition is a problem for other components using the eventlog to track the job state.   I think we'll fix that by submitting a PR to RFC 21 to say that the `priority` event is persistent and add a new event `restart` or `prority-invalidate`, posted on instance restart, that transitions SCHED to PRIORITY.

2) We could rename the event for updates to `priority-update` and make it a JOURNAL_ONLY event.  However, if we restart job-info, the journal may have wrapped so that the latest priority values are unavailable, and job-info would remain out of date until priorities are updated.

(hopefully that captures the problem)

**Comment by garlick** [comment]
@chu11 idea

> Could the job-manager be queried with a set of jobids and the queue priorities could be returned? Or perhaps it simply informs the job-manager to write those queue priorities to the journal and doesn't even have to respond to the RPC?

@garlick idea

> I was just wondering if we should get rid of the JOURNAL_ONLY event idea and instead have a streaming RPC that tracks requested non-persistent attributes like priority?

**Comment by chu11** [comment]
> You'd know better than I what complications would arise when trying to consume data from both, when they are only loosely synchronized. I guess you'd have to deal with a priority update before you know about the job from the journal. Maybe if those things were kept in a separate hash in job-info that's accessed on demand (rather than having an RPC in the job-manager to provide them on demand?) Could we go one further and put that "attribute hash" or priority hash or whatever in a reusable libjob container?

First thought is if we have 2 streaming RPCs, and neither block packets on the job-manager KVS batch commit, would packets arrive in order at the receiver, no matter the order (or which RPC) sent on the sender side?  I think so?  The KVS commit blocking sends from before is the key difference here.

Other thought, with annotations, they are sort of "best effort" in job-info, so raciness to some degree we care less about it.  For example, scheduler annotations shouldn't change after the job is running.  If the annotation arrives late after a job enters the RUN state, do we care?

Similar could be said about the queue priority.  Its predominant purpose is to order the jobs being listed.  Does it matter if that's out of sync for awhile?

**Comment by garlick** [comment]
So job-info would 1) replay event logs, 2) start consuming watching journal, 3) start this new RPC for getting initial, then updated priorities.  No races there?

**Comment by chu11** [comment]
> So job-info would 1) replay event logs, 2) start consuming watching journal, 3) start this new RPC for getting initial, then updated priorities. No races there?

I don't think any races .. could be famous last words once I dig into it.  But it minimally will not have the same race as the old races.

**Comment by grondo** [comment]
I'm not following this discussion too closely, but just wanted us to keep in mind whether we want to open up the job-manager info protocol to other users in the future. And if so, if the decisions being made here would make it onerous for developers to get job information directly from the job-manager.

As an example use case to consider, we had talked about a python-based job notification service which perhaps could use the job-manager journal. If some information is not available from the journal, this could be confusing.  Now, a job notification service may not need to get updates of job priority, but if there is something special about priority that requires it to be handled differently from all other job information, will there perhaps be some other special job attributes in the future that would need yet another way to register for updates?

The complexity could always be hidden behind a library, but all I'm asking is if there is some way to embed the extra complexity in the "journal" interface instead. E.g. could updates of ephemeral job attributes like priority still be sent to consumers via the journal interface, but not be persisted in the in-memory journal? Would that be the same amount of code in the job-manager, but a lot less code in clients, and more extensible to future dynamic attributes and more clear?

I apologize if I'm way off the mark or missed an obvious reason this can't be done.

**Comment by garlick** [comment]
The model for the journal is for a consumer to 1) walk all the jobs in the KVS reading their eventlogs, then 2) apply the journal events to keep up to date.  The problem with the JOURNAL_ONLY events is that, once the journal wraps, the info is lost to consumers because it cannot be initialized from the KVS.

Simple example: job manager re-prioritizes 10,0000 jobs on restart, emitting JOURNAL_ONLY events, but the journal wraps at 1,000 entries.  job-info loads a few seconds later, establishes the RPC for the journal, then walking the KVS.    job-info never sees the 9,000 events that wrapped, because they were not committed to the KVS eventlogs.

One thing we could do that might make the journal more generally useful would be to retain the eventlogs in the job manager's memory.   Then we could make the full event stream (from the beginning) available for all active jobs.  job-info, and other consumers, wouldn't have to walk the full KVS job directory, throwing out inactive jobs, as a bonus.

As we've discussed before, I probably paid too much attention to the memory footprint in the original job manager design, and usability has suffered as a result.  Considering that most jobs only accrue a dozen or so events during their life cycle, and we can throw them out when they transition to INACTIVE, this is probably not a huge overhead.

**Comment by garlick** [comment]
I'm feeling unenlightened right now and don't want to just toss off a half-baked design that we'll have to redo later.  What if we just say for now that we post a `priority-update` event when the priority changes, and _do_ store it in the KVS eventlog?  Then job-info et al can track it like it tracks everything else now.  We can also let the priority plugin post useful info in the event context, such as factor values and weights, which will aid in debugging and transparency while we are working on the plugins.

Then we should give ourselves some time to ponder how to do this properly.  I think we should try to shoot for a unified job-manager interface for tracking job data that could be used by job-info as well as other components, and does not require the user to traverse the entire KVS job directory to get the initial values.

**Comment by chu11** [comment]
@garlick that seems like a good plan for now, so we can move forward

**Comment by chu11** [comment]
per a discussion in #3381 (splitting it off into here), just brainstorming as i prototyped this morning

- there shouldn't be a default priority (formerly "queue priority")
- however, in job-info, pending jobs should be sorted by queue priority first.
- So what to do if the priority is not yet set (such as the job is in state DEPEND or PRIORITY state).
  - in #3381 suggestion is to sort by jobid, which seems reasonable
  - but that ultimately means we have to pick if DEPEND/PRIORITY state jobs should be listed before/after jobs in SCHED state.  I'm not sure if there's a preference, but it would make more sense to me that they are listed after all jobs in SCHED state jobs.

  - thought 1- well, if they are listed after all jobs in SCHED state, perhaps it would make sense to just assign them a default priority of 0? i.e. the lowest priority?  Then the "sort" on the pending list just works out.

  - thought 2 - perhaps the "pending list" should be logically split into jobs with a queue priority and jobs without a queue priority.  It would always be presented to users as a single list of results.

**Comment by chu11** [comment]
other thing that came up in prototyping:

```
/* zlistx_insert() and zlistx_reorder() take a 'low_value' parameter
 * which indicates which end of the list to search from.
 * false=search begins at tail (lowest priority, youngest)
 * true=search begins at head (highest priority, oldest)
 * Attempt to minimize search distance based on job priority.
 */
```

without a default priority and unknown how priority plugins will assign priorities, there's no way to calculate a search direction with a hard coded default.  Some "running average" of the current queue priorities would be best.

But as we discussed shortly in #3052, a list data structure may not be the best long term data structure for regularly changing priorities of jobs.  So short term we could perhaps just hardcode `PRIORITY_MAX / 2` as the middle number for a search, assuming that the searches aren't eating up too many CPU cycles.  And if it is eating up CPU, we should revisit (minimally with a running average calculated average, or debate a better data structure).

**Comment by garlick** [comment]
> thought 1- well, if they are listed after all jobs in SCHED state, perhaps it would make sense to just assign them a default priority of 0? i.e. the lowest priority? Then the "sort" on the pending list just works out.

We're just talking about how you internally represent these in `job-info` so they can be retrieved in order for `flux jobs` right?  Sure, this seems fine.  If jobs are listed in ascending priority order, then it kind of makes sense that jobs that are not schedulable just get listed before those.  And if you're are breaking ties with the FLUID then _those_ will be listed in submit order, which sounds fine.

On the zlistx question, probably not something to agonize over right now IMHO.

**Comment by chu11** [comment]
> We're just talking about how you internally represent these in job-info so they can be retrieved in order for flux jobs right? 

correct.  

> Sure, this seems fine. If jobs are listed in ascending priority order, then it kind of makes sense that jobs that are not schedulable just get listed before those. And if you're are breaking ties with the FLUID then those will be listed in submit order, which sounds fine.

Yes, although I think you mean descending?  highest priority jobs listed first.

**Comment by chu11** [comment]
aww crud, this issue is closed via commit 902af08a5cbe6c4163df5a846cd2e83862ab77ba in PR #3404 

I typoed the "Fixes" line in the commit.
